frame/evm/src/runner/stack.rs

```diff
@@ -827,6 +910,222 @@ where
+			proof_size_limit
+		} else {
+			return Ok(());
+		};
+
+		let maybe_record = match gas_cost {
+			GasCost::ExtCodeSize { target_is_cold }
+			| GasCost::Balance { target_is_cold }
+			| GasCost::ExtCodeHash { target_is_cold }
+			| GasCost::CallCode { target_is_cold, .. }
+			| GasCost::StaticCall { target_is_cold, .. }
+			| GasCost::ExtCodeCopy { target_is_cold, .. }
+			| GasCost::SLoad { target_is_cold, .. }
+			| GasCost::DelegateCall { target_is_cold, .. }
+			| GasCost::SStore { target_is_cold, .. }
+			| GasCost::Call { target_is_cold, .. } => target_is_cold,
+
+			GasCost::Create => true,
+
+			GasCost::Suicide {
+				target_is_cold,
+				already_removed,
+				..
+			} => target_is_cold && !already_removed,
+
+			_ => false,
+		};
+
+		if !maybe_record {
+			return Ok(());
+		}
+		// Proof size is fixed length for writes (a 32-byte hash in a merkle trie), and
+		// the full key/value for reads. For read and writes over the same storage, the full value
+		// is included.
+		// For cold reads involving code (call, callcode, staticcall and delegatecall):
+		//	- We depend on https://github.com/paritytech/frontier/pull/893
+		//	- Try to get the cached size or compute it on the fly
+		//	- We record the actual size after caching, refunding the difference between it and the initially deducted
+		//	contract size limit.
+		let opcode_proof_size = match opcode {
+			// Basic account fixed length
+			Opcode::BALANCE => U256::from(ACCOUNT_BASIC_PROOF_SIZE),
+			// TODO requires https://github.com/paritytech/frontier/pull/893
+			Opcode::EXTCODESIZE
+			| Opcode::EXTCODECOPY
+			| Opcode::EXTCODEHASH
+			| Opcode::CALLCODE
+			| Opcode::CALL
+			| Opcode::DELEGATECALL
+			| Opcode::STATICCALL => {
+				let address = match target {
+					StorageTarget::Address(address) | StorageTarget::Slot(address, _) => address,
+					// This must be unreachable, a valid Target must be set for this opcode(s)
+					// TODO decide how do we want to gracefully handle.
+					_ => return Err(ExitError::OutOfGas),
+				};
+				// First try to record fixed sized `AccountCodesMetadata` read
+				// Temptatively 20 + 8 + 32
+				weight_info.try_record_proof_size_or_fail(ACCOUNT_CODES_METADATA_PROOF_SIZE)?;
+				let size = if let Some(meta) = <AccountCodesMetadata<T>>::get(address) {
+					meta.size
+				} else {
+					// If it does not exist, try to record `create_contract_limit` first.
+					weight_info.try_record_proof_size_or_fail(size_limit)?;
+					let meta = Pallet::<T>::account_code_metadata(address);
+					let actual_size = meta.size;
+					// Refund if applies
+					weight_info.refund_proof_size(size_limit.saturating_sub(actual_size));
+					actual_size
+				};
+				U256::from(size)
+			}
+			// (H160, H256) double map blake2 128 concat key size (68) + value 32
+			Opcode::SLOAD => U256::from(ACCOUNT_STORAGE_PROOF_SIZE),
+			// Fixed trie 32 byte hash
+			Opcode::SSTORE | Opcode::CREATE | Opcode::CREATE2 | Opcode::SUICIDE => {
```

Comment:
This is wrong for chains with state version v0.
The case `Opcode::SUICIDE ` should be handled separatly, depending of the state version. If the state version is v0, it should account for the contractcode size

Reply:
pushed